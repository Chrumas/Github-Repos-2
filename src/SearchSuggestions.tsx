import React from "react";
import { useState, useEffect, useImperativeHandle, useRef } from "react";
import {
  useSearchContext,
  GlobalSearch,
  useUserUpdateContext,
  User,
} from "./ContextProvider";
import { StyledSearchSuggestions } from "./styles/SearchSuggestions.styled";
import { ErrorMessage } from "./ErrorMessage";

//This component is resposible for viewing the search suggestions (data from github api search engine). We choose a user by clicking on him/her.
export const SearchSuggestions = React.forwardRef<any>((props, myref) => {
  const search: GlobalSearch = useSearchContext();
  const suggestionsRefs = useRef<HTMLDivElement[]>([]);
  const [searchTerm, setSearchTerm] = useState<GlobalSearch>({ search: "" });
  const [searchSuggestionTable, setTable] = useState<any>("");
  const [errorMessage, setErrorMessage] = useState("");
  const setUserContext = useUserUpdateContext();

  useImperativeHandle(myref, () => ({
    refCallback: suggestionsRefs,
  }));

  useEffect(() => {
    setSearchTerm(search);
  }, [search]);

  //Fetches data about searched users
  useEffect(() => {
    if (searchTerm !== { search: "" } || searchTerm !== undefined) {
      let uri = "https://api.github.com/search/users?q=" + searchTerm;
      try {
        fetch(uri)
          .then((response) => {
            if (!response.ok) {
              if (response.status === 403) {
                setErrorMessage("API limit exceeded!");
                throw new Error("API limit exceeded!");
              }
            }
            return response.json();
          })
          .then((data) => {
            setErrorMessage("");
            var map = data.items.map((choice: any, index: number) => {
              let user: User = {
                name: choice.login,
                avatarUrl: choice.avatar_url,
                githubLink: choice.html_url,
              };
              return (
                <div
                  data-testid="suggestion"
                  key={index}
                  className="search-suggestion"
                  ref={(element) => {
                    if (element !== null)
                      suggestionsRefs.current[index] = element;
                  }}
                  onClick={() => {
                    setUserContext(user);
                  }}
                >
                  <div className="mini-avatar"></div>
                  {choice.login}
                </div>
              );
            });
            setTable(map);
          })
          .catch((e) => {
            console.log(e);
          });
      } catch {}
    }
  }, [searchTerm, setUserContext]);

  return (
    <StyledSearchSuggestions>
      <ErrorMessage inputColor="red" inputName={errorMessage} />
      <div className="search-suggestions-map-wrapper">
        <div className="search-suggestions-map">{searchSuggestionTable}</div>
      </div>
    </StyledSearchSuggestions>
  );
});

export default SearchSuggestions;
