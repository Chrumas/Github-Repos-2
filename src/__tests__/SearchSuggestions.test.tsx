import * as React from "react";
import * as ReactDOM from "react-dom";
import { SearchSuggestions } from "../SearchSuggestions";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<SearchSuggestions />, div);
});

it("matches snapshot", () => {
  const tree = renderer.create(<SearchSuggestions />).toJSON();
  expect(tree).toMatchSnapshot();
});
