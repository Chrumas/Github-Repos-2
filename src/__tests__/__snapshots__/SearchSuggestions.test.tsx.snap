// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="sc-bdvvtL hZTdAJ"
>
  <div
    className="search-suggestions-map-wrapper"
  >
    <div
      className="search-suggestions-map"
    >
      
    </div>
  </div>
</div>
`;
